File: androidApp/src/main/java/com/cas/musicplayer/firebase/Storage.kt
package com.cas.musicplayer.firebase

import android.util.Log
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageException
import com.mousiki.shared.fs.PathComponent
import com.mousiki.shared.utils.ConnectivityChecker
import com.mousiki.shared.utils.StorageApi
import java.io.File
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine


class Storage(
    private val firebaseStorage: FirebaseStorage
) : StorageApi {

    override suspend fun downloadFile(
        remoteUrl: String,
        path: PathComponent,
        connectivityState: ConnectivityChecker,
        logErrorMessage: String
    ): PathComponent {
        val file = File(path.component!!)
        val localFile = downloadFileFormFirebaseStorage(
            remoteUrl = remoteUrl,
            localFile = file,
            connectivityState = connectivityState,
            logErrorMessage = logErrorMessage
        )
        return PathComponent(localFile.absolutePath)
    }

    private suspend fun downloadFileFormFirebaseStorage(
        remoteUrl: String,
        localFile: File,
        connectivityState: ConnectivityChecker,
        logErrorMessage: String = "Cannot load $remoteUrl file from firebase"
    ): File {
        if (!localFile.exists()) {
            val connectedBeforeCall = connectivityState.isConnected()
            var retryCount = 0
            var fileDownloaded = false
            var fileExist = true
            while (retryCount < MAX_RETRY_FIREBASE_STORAGE && !fileDownloaded && fileExist) {
                retryCount++
                fileDownloaded = suspendCoroutine { continuation ->
                    val ref = firebaseStorage.getReferenceFromUrl(remoteUrl)
                    Log.d("StorageApi", "Calling $remoteUrl")
                    ref.getFile(localFile).addOnSuccessListener {
                        Log.d("StorageApi", "Success $remoteUrl")
                        continuation.resume(true)
                    }.addOnFailureListener {
                        Log.d("StorageApi", "Failure $remoteUrl with exception $it")
                        if ((it as? StorageException)?.errorCode == StorageException.ERROR_OBJECT_NOT_FOUND) {
                            fileExist = false
                        }
                        continuation.resume(false)
                    }
                }
            }
            if (!fileDownloaded) {
                // Log error
                FirebaseCrashlytics.getInstance().log(
                    "$logErrorMessage ==> after $retryCount retries," +
                            "\n Is Connected before call: $connectedBeforeCall" +
                            "\n Is Connected after call:${connectivityState.isConnected()}"
                )
            }
        }
        return localFile
    }
}

private val MAX_RETRY_FIREBASE_STORAGE = 4

File: iosApp/Mousiki/Common/Firebase/Storage.swift
//
//  Storage.swift
//  Mousiki
//
//  Created by A.Yabahddou on 2/1/21.
//  Copyright © 2021 YabaSoft. All rights reserved.
//

import Foundation
import FirebaseStorage
import shared

class MousikiAppStorage:  StorageApi {
    
    fileprivate let firebaseStorage = Storage.storage()
    
    func downloadFile(
        remoteUrl: String,
        path: PathComponent,
        connectivityState: ConnectivityChecker,
        logErrorMessage: String,
        completionHandler: @escaping (PathComponent?, Error?) -> Void) {
        
        print("\n\n")
        print("###################")
        print("Call to load file from firebase remote=\(remoteUrl) and to local=\(path.component!)")
        print("###################")
        print("\n\n")
        
        print("Thread is \(Thread.current)")
        let reference = self.firebaseStorage.reference(forURL: remoteUrl)
        
        let fileUrl = URL(string: "file://\(path.component!)")
        reference.write(toFile: fileUrl!) { (url, error) in
            print("Response Thread is \(Thread.current)")
            print("\n\n")
            print("###################")
            print("Got storage response from \(remoteUrl) and error=\(error)")
            print("###################")
            print("\n\n")
            completionHandler(path, nil)
        }
    }
    
}
